test/e2e/ingress_utils.go

```diff
@@ -151,14 +281,413 @@ func createSecret(kubeClient *client.Client, ing *extensions.Ingress) (host stri
 		},
 	}
 	var s *api.Secret
-	if s, err = kubeClient.Secrets(ing.Namespace).Get(tls.SecretName); err == nil {
+
+	if err = client.Get().Namespace(ing.Namespace).Resource(secretResourceName).Name(tls.SecretName).Do().Into(s).Error(); err == nil {
 		// TODO: Retry the update. We don't really expect anything to conflict though.
 		framework.Logf("Updating secret %v in ns %v with hosts %v for ingress %v", secret.Name, secret.Namespace, host, ing.Name)
 		s.Data = secret.Data
-		_, err = kubeClient.Secrets(ing.Namespace).Update(s)
+		err = client.Put().Namespace(ing.Namespace).Resource(secretResourceName).Name(tls.SecretName).Body(s).Do().Error()
 	} else {
 		framework.Logf("Creating secret %v in ns %v with hosts %v for ingress %v", secret.Name, secret.Namespace, host, ing.Name)
-		_, err = kubeClient.Secrets(ing.Namespace).Create(secret)
+		err = client.Post().Namespace(ing.Namespace).Resource(secretResourceName).Name(tls.SecretName).Body(secret).Do().Error()
 	}
 	return host, cert, key, err
 }
+
+func describeIng(ns string) {
+	framework.Logf("\nOutput of kubectl describe ing:\n")
+	desc, _ := framework.RunKubectl(
+		"describe", "ing", fmt.Sprintf("--namespace=%v", ns))
+	framework.Logf(desc)
+}
+
+func cleanupGCE(gceController *GCEIngressController) {
+	if pollErr := wait.Poll(5*time.Second, lbCleanupTimeout, func() (bool, error) {
+		if err := gceController.Cleanup(false); err != nil {
+			framework.Logf("Still waiting for glbc to cleanup: %v", err)
+			return false, nil
+		}
+		return true, nil
+	}); pollErr != nil {
+		if cleanupErr := gceController.Cleanup(true); cleanupErr != nil {
+			framework.Logf("WARNING: Failed to cleanup resources %v", cleanupErr)
+		}
+		framework.Failf("Failed to cleanup GCE L7 resources.")
+	}
+}
+
+// Cleanup cleans up cloud resources.
+// If del is false, it simply reports existing resources without deleting them.
+// It always deletes resources created through it's methods, like staticIP, even
+// if del is false.
+func (cont *GCEIngressController) Cleanup(del bool) error {
```

Comment:
This is a very long function.  Would it make sense to break it up into smaller sub-functions?